### DISCUSSION TITLE
Is O3DE DX12_2 ready to ship a game on console?

### DISCUSSION THREAD

**Discussion by Meshingaround** [o3de/o3de#16397] [Q&A] [answered]
Just as above is the engine ready to ship a game using dx12_2? How does the open nature work with closed Devkit stuff legally? Can dx12_2 features like directstorage & sampler feedback(streaming) be added to it legally? Technically? What's the word on mesh shaders & 64bit intrinsic support? Has someone written a viable DXR ray tracer?

**Comment by AMZN-daimini** [comment]
@o3de/sig-graphics-audio

**Comment by moudgils** [comment, marked as answer]
The engine has a DX12 gem which implements the DX12 backend. It was built with steam platforms in mind (i.e PC and Linux). It currently has support for DXR ray tracing API amongst other features. It does not have support for mesh shading as we dont have a specific use case where mesh shading will provide a significant improvement yet. Plus when we add mesh shading support we will need to think of other backends like Vulkan and Metal in order to add correct interface at higher RHI level in order to ensure the capability is supported across all platforms. Other new DX12 features like directstorage and sampler feedback are currently not supporeted by the DX12 gem but can be added by anyone if they need it. In this case you would work with graphics sig, create a RFC technical document on how to add these features and then in the end put up PRs with the implementation. 64bit intrinsic support was recently added for linux but not pc although I would imagine it should not be too hard to extend to pc now. 

Console implementation - Given legal implications of console backend we have added support to PAL-ify (PAL = platform abstraction layer) all gems in a manner where some code would be added in a restricted manner. So there is code layout that allows us to extend a gem's functionality where the new functionality for platforms like xbox can reside on a restricted repo. You will need to work with Linux foundation to get access to that repo and I would imagine one of the requirement would be to be part of Microsoft dev group. When working on O3de's renderer we had done some initial work to get basic rendering working on Xbox consoles but have not done much work thereafter. For example all the Xbox specific dx12 api related to swap chain management resides in this extended functionality. In order to ship a game on console I would imagine more work is need to ensure better performance and address any bugs that may have crept up since the initial work was done.

**Comment by Meshingaround** [comment]
Thank you for your response. I'm not the best on the subject of rendering as I'm a newbie to this side...however,  when you say there's not a specific use case for dx12 mesh shaders what about tasks, LOD & tessellation?

I'm more interested in utilizing a SF & mesh shader for procedural generation and massive parallelization like voxel grid generation for SVOGI against heavy poly's with RT HW for indirect & optimization. So if I needed to render a complex forest I'm limited to a standard pipeline with many draw calls? How does that work exactly?

Again, thank you for your response.

________________________________

**Comment by moudgils** [comment]
All your examples above would be considered good use cases for mesh shaders but we (I can only speak for my team) have not needed to add support for them as yet. The GI solution for O3de use RTX-GI and as a result there was no need for SVOGI. Tessellation support is heavily content dependent as yet with the type of content we have built we have had no need to add tessellation support. 

To be honest the only use case that will make a good use case for mesh shading is optimizing an Indirect rendering pipeline. The renderer at the moment uses Forward+ rendering pipeline with some hybrid G-buffers. Eventually we think that building a fully GPU driven pipeline as an alternative pipeline is a good idea especially considering the decoupled nature of the renderer where the pipeline is data driven and materials are decoupled from lighting. After building an indirect pipeline it would make sense to build out support for mesh shaders as it could provide opportunities to improve perf further. It is simply a case of where to spend engineering cycles. 

Given the open source modular nature of this engine anyone is able to build out this tech so it is not beholden on one team to do this work. Other teams/engineers are more than welcome to contribute their time. Adding mesh shading support at RHI level has actually become a bit easier as apple has released support to compile metal byte code from dxil directly. I wrote a RFC on how we could use this tool for the metal backend (https://github.com/o3de/sig-graphics-audio/blob/main/rfcs/AppleShaderConverterIntegration/AppleShaderConverterIntegration.md). This means that we should be able to add an interface at RHI level which can be easily extended by DX12, Vk and metal backends. Once that is done a rendering pipeline (i.e Indirect pipeline) at higher level (i.e RPI gem) will be able to utilize mesh shaders for their purposes.
